blockchains/erc20/lib/timestamps_cache.js

```diff
@@ -1,45 +1,54 @@
 'use strict';
-const DATA_MISSING = -1;
 
 class TimestampsCache {
-  constructor() {
+  constructor(ethClient, fromBlock, toBlock) {
     this.timestampStore = {};
-  }
+    this.rangeSize = toBlock - fromBlock + 1;
+
+    const blockRequests = [];
+    for (let i = fromBlock; i <= toBlock; i++) {
+      blockRequests.push(
+        ethClient.request(
+          'eth_getBlockByNumber',
+          [i, false],
+          undefined,
+          false
+        )
+      );
+    }
 
-  saveTimestampInStore(blockNumber, timestamp) {
-    this.timestampStore[blockNumber] = timestamp;
+    this.responsePromise = ethClient.request(blockRequests);
   }
 
-  getTimestampFromStore(blockNumber) {
-    if (Object.prototype.hasOwnProperty.call(this.timestampStore, blockNumber)) {
-      return this.timestampStore[blockNumber];
+  async waitResponse(web3Wrapper) {
+    const resultsArray = await this.responsePromise;
+    if (!Array.isArray(resultsArray)) {
+      throw new Error('Blocks response is not an array');
+    }
+    if (resultsArray.length !== this.rangeSize) {
+      throw new Error(`Expected ${this.rangeSize} but got ${resultsArray.length} blocks response`);
     }
 
-    return DATA_MISSING;
-  }
+    for (const result of resultsArray) {
```

Comment:
There are libraries like underscore.js (old one) which can bring style here. `map, filter, fold` guys are waiting to improve readability

Reply:
Nowadays functional appraoch is available through the language itself, ex:

```
const newTimestampStore = resultsArray.reduce((timestampStore, result) => {
  const blockNumber = web3Wrapper.parseHexToNumber(result.result.number);
  const blockTimestamp = web3Wrapper.parseHexToNumber(result.result.timestamp);
  return {
    ...timestampStore,
    [blockNumber]: blockTimestamp
  };
}, {});
```
However even if we presume that this is better as there are no 'side effects' the performance is worse. The spred operator would construct an extra object on each iteration. Also spread operator would throw SO exception on big objects.